llvm/lib/Transforms/IPO/AttributorAttributes.cpp

```diff
@@ -12509,6 +12509,34 @@ struct AAIndirectCallInfoCallSite : public AAIndirectCallInfo {
 
 /// ------------------------ Address Space  ------------------------------------
 namespace {
+
+template <typename InstType>
+static bool makeChange(Attributor &A, InstType *MemInst, const Use &U,
+                       Value *OriginalValue, PointerType *NewPtrTy,
+                       bool UseOriginalValue) {
+  if (U.getOperandNo() != InstType::getPointerOperandIndex())
+    return false;
+
+  auto *TTI = A.getInfoCache().getAnalysisResultForFunction<TargetIRAnalysis>(
+      *MemInst->getFunction());
+  if (!TTI)
+    return false;
+
+  unsigned OldAS = MemInst->getPointerAddressSpace();
+  if (MemInst->isVolatile() && !TTI->hasVolatileVariant(MemInst, OldAS))
```

Comment:
For `hasVolatileVariant`, the reading seems like it needs to pass new AS to see there is the corresponding volatile variant, but in `InferAddressSpaces` it uses the old AS.

Reply:
That seems wrong. Can we check hasVolatileVariant? It sounds like it would check if there is a variant in the new AS, why do we care about the old one?

Reply:
I think `hasVolatileVariant` indeed needs to pass the new AS instead of the old one based on the comment. Also, it returns `false` by default, which prevents any AS inference.